### PR TITLE
872 as html fix@main

### LINKED ISSUE #872: update `as_html` to accept "\n" reformatting
https://github.com/insightsengineering/rtables/blob/bdafa9d155ef71c7bc829d17c5ba78ae5a7c8308/R/as_html.R#L85

initialization uses the matrix object which does not account for new lines from any text wrapping using "\n". 
replace the calculation with the "string matrix"

### PR DESCRIPTION
**Pull request by ayogasekaram**
closes #872

### CODE REVIEW

**Review by edelarua — APPROVED** [review]
Lgtm!

File: R/as_html.R [hunk 1]
```diff
@@ -73,23 +73,24 @@ as_html <- function(x,
   if (is.null(x)) {
     return(tags$p("Empty Table"))
   }
-
+
```
Comment (Melkiades) [inline R/as_html.R:76]: I think you need to remove these

File: R/as_html.R [hunk 2]
```diff
@@ -102,21 +103,25 @@ as_html <- function(x,
...
-
+  
   # row labels style
-  for (i in seq_len(nrow(x))) {
-    indent <- mat$row_info$indent[i]
-    if (indent > 0) { # indentation
-      cells[i + nlh, 1][[1]] <- htmltools::tagAppendAttributes(cells[i + nlh, 1][[1]],
-        style = paste0("padding-left: ", indent * 3, "ch;")
-      )
+  for (i in 1:nr) {
```
Comment (Melkiades) [inline R/as_html.R:115]:
```suggestion
  for (i in seq_len(nr)) {
```

File: R/as_html.R [hunk 3]
```diff
@@ -125,81 +130,81 @@ as_html <- function(x,
...
       ) # nocov
     }
-
-    for (ii in unique(mat$line_grouping)) {
-      rows <- which(mat$line_grouping == ii)
-      should_display_col <- apply(mat$display[rows, , drop = FALSE], 2, any)
+    
+    for (ii in unique(mf_lgrouping(mat))) {
+      rows <- which(mf_lgrouping(mat) == ii)
+      should_display_col <- apply(mf_display(mat)[rows, , drop = FALSE], 2, any)
```
Comment (Melkiades) [inline R/as_html.R:175]: Thanks Abi!! this was really needed (also everywhere else in principles)

**Review by edelarua — CHANGES_REQUESTED** [review]
Hi @ayogasekaram,

The row labels style section you removed isn't currently used in teal apps but is still useful (and could be implemented there in the future).

For example, when the following code is run, see the difference in html output with/without this section included:

```r
tbl <- basic_table() %>%
  split_rows_by("BMRKR2") %>%
  analyze("AGE") %>%
  build_table(ex_adsl)

Viewer(tbl, bold = c("header", "row_names"))
```

Produces without this section:
Screenshot 2024-06-07 154729

Produces with this section:
Screenshot 2024-06-07 154744

I would add the removed code back in (and maybe another example like the one I gave above?) and then I think you're good to go. That should also fix the failing test :)

Thanks!!

**Comment by edelarua** [comment]
Sorry, I forgot to also mention that the indentation code (each indent is tripled) was requested at one point to improve visibility within teal modules.

For example,

Without indentation:
  
  Screenshot 2024-06-07 172520

With indentation:
  
  Screenshot 2024-06-07 172359

**Comment by shajoezhu** [comment]
hi @ayogasekaram , could you raise PRs downstream, and see how much downstream breaking changes there will be. thanks!

- [ ] tern
- [ ] tern.rbmi
- [ ] tern.gee
- [ ] tern.mmrm
- [ ] teal.modules.clinical
- [ ] scda.test

Thanks a lot!

**Comment by ayogasekaram** [comment]
> hi @ayogasekaram , could you raise PRs downstream, and see how much downstream breaking changes there will be. thanks!
> 
> * [ ]  tern
> * [ ]  tern.rbmi
> * [ ]  tern.gee
> * [ ]  tern.mmrm
> * [ ]  teal.modules.clinical
> * [ ]  scda.test
> 
> Thanks a lot!

Hey @shajoezhu! Thanks for taking a look. There shouldn't be downstream changes in any of those packages as they don't call the `as_html` function. The function also should work exactly as before - it just accommodates the newline character if present so there aren't any tables that were modified. I ran the checks for each of those packages locally with this rtables change and they've all passed but I can raise the PRs just in case!

**Comment by shajoezhu** [comment]
@ayogasekaram I see. I saw an indentation change, and thought that might be breaking changes. can you give it a try for scda.test, if everything is ok, we can quickly move on, and include this in the coming release. thanks!

**Comment by ayogasekaram** [comment]
@shajoezhu No errors in scda.test :)

File: NEWS.md [hunk 4]
```diff
@@ -21,6 +21,7 @@
  * Fixed bug in `as_html` preventing indentation from being applied in `Viewer` output.
  * `col_counts<-` and `col_total<-` methods now explicitly convert `value` to integer, by @gmbecker.
  * `col_gap` is now respected in `nlines` row methods, and thus by `make_row_df`, by @gmbecker.
+ * update `as_html` to accommodate `\n` characters.
```
Comment (edelarua) [inline NEWS.md:24]: Past tense
```suggestion
 * Updated `as_html` to accommodate `\n` characters.
```

File: R/as_html.R [hunk 5]
```diff
@@ -80,15 +80,16 @@ as_html <- function(x,
 
   nlh <- mf_nlheader(mat)
   nc <- ncol(x) + 1
+  nr <- length(mf_lgrouping(mat))
 
   # Structure is a list of lists with rows (one for each line grouping) and cols as dimensions
-  cells <- matrix(rep(list(list()), (nlh + nrow(x)) * (nc)), ncol = nc)
+  cells <- matrix(rep(list(list()), (nr) * (nc)), ncol = nc)
```
Comment (edelarua) [inline R/as_html.R:86]: Don't really need the extra brackets
```suggestion
  cells <- matrix(rep(list(list()), nr * nc), ncol = nc)
```

File: R/as_html.R [hunk 6]
```diff
@@ -110,32 +111,36 @@ as_html <- function(x,
...
-  # label rows style
-  if ("label_rows" %in% bold) {
-    which_lbl_rows <- which(mat$row_info$node_class == "LabelRow")
-    cells[which_lbl_rows + nlh, ] <- lapply(
-      cells[which_lbl_rows + nlh, ],
-      htmltools::tagAppendAttributes,
-      style = "font-weight: bold;"
-    )
-  }
-
```
Comment (edelarua) [inline R/as_html.R:138]: Why was this section removed? I don't think it's related to the changes for `\n` since it only controls bolding for label rows

**Review by edelarua — COMMENTED** [review]
Hi Abi! Just a few small comments.

**Review by edelarua — APPROVED** [review]
Lgtm - thanks Abi!!

**Review by Melkiades — APPROVED** [review]
Lgtm!! Thanks Abi :) wonderful work

**Pull request merged**